### PR TITLE
Opt: TargetEval values in SolutionExport

### LINKED ISSUE #1625: [TASK] Additional TargetEvaluation variables in SolutionExport for Optimizer
--------
Issue Description
--------

It'd be nice if values from the TargetEvaluation could be added to the SolutionExport for the Optimizer.

All that would be needed is modifying `RavenSampled` `_updateSolutionExport` to check for additional variables names that `self._solutionExport` wants, and are in `rlz`, but haven't been added yet.

We tested this with @SamKerber and it seemed to work as expected, and didn't slow the query down much.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1625

##### What are the significant changes in functionality due to this change request?
- Allows including any variables in the `TargetEvaluation` in the `SolutionExport` for `RavenSampled` Optimizers
- Improves errors related to failed model runs in `RavenSampled` Optimizers

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Optimizers/RavenSampled.py [hunk 1]
```diff
@@ -332,7 +332,12 @@ def finalizeSampler(self, failedRuns):
         bestTraj = traj
         bestValue = val
     # further check active unfinished trajectories
-    traj = 0
+    ## FIXME why should there be any active, unfinished trajectories when we're cleaning up sampler?
+    traj = 0 # FIXME why only 0?? what if it's other trajectories that are active and unfinished?
```
Comment (PaulTalbot-INL) [inline framework/Optimizers/RavenSampled.py:336]: The two FIXME notes here aren't part of this PR; they're comments on previously-added code, FYI.

File: framework/Optimizers/RavenSampled.py [hunk 2]
```diff
@@ -599,6 +604,10 @@ def _updateSolutionExport(self, traj, rlz, acceptable, rejectReason):
     if self._minMax == 'max':
       objValue *= -1
     toExport[self._objectiveVar] = objValue
+    # check for anything else that solution export wants that rlz might provide
+    for var in self._solutionExport.getVars():
+      if var not in toExport and var in rlz:
+        toExport[var] = rlz[var]
```
Comment (wangcj05) [inline framework/Optimizers/RavenSampled.py:610]: Should these lines be moved to line 614 (just after the updates of toExport dict)?

File: tests/framework/AnalyticModels/optimizing/beale.py [hunk 3]
```diff
@@ -35,6 +35,8 @@ def run(self,Inputs):
     @ Out, None
   """
   self.ans = evaluate(self.x,self.y)
+  # test secondary variable passthrough
+  self.aux_ans = self.x + self.y
```
Comment (wangcj05) [inline tests/framework/AnalyticModels/optimizing/beale.py:39]: please change "aux_ans" to "auxAns".

File: tests/framework/Optimizers/GradientDescent/basic.xml [hunk 4]
```diff
@@ -92,19 +92,19 @@
 
   <Models>
     <ExternalModel ModuleToLoad="../../../framework/AnalyticModels/optimizing/beale" name="beale" subType="">
-      <variables>x,y,ans</variables>
+      <variables>x,y,ans,aux_ans</variables>
```
Comment (wangcj05) [inline tests/framework/Optimizers/GradientDescent/basic.xml:95]: please update the "aux_ans" here also

File: tests/framework/Optimizers/GradientDescent/gold/Basic/opt_export.xml [hunk 5]
```diff
@@ -2,7 +2,7 @@
   <DataSet type="Static">
     <general>
       <inputs>trajID</inputs>
-      <outputs>x,y,func,const,ans,stepSize,iteration,accepted,conv_gradient</outputs>
+      <outputs>x,y,func,const,ans,aux_ans,stepSize,iteration,accepted,conv_gradient</outputs>
```
Comment (wangcj05) [inline tests/framework/Optimizers/GradientDescent/gold/Basic/opt_export.xml:5]: Do you need this file? It seems to me this file can be removed.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
I have some minor comments for your consideration.

File: tests/framework/AnalyticModels/optimizing/beale.py (hunk 3 above)
Comment (PaulTalbot-INL) [reply tests/framework/AnalyticModels/optimizing/beale.py:39]: this isn't a RAVEN variable, I don't think that applies here. We haven't applied code standards to ExternalModels at any point, I don't think ... maybe we need an issue for updating all external models in the repository.

File: tests/framework/Optimizers/GradientDescent/gold/Basic/opt_export.xml (hunk 5 above)
Comment (PaulTalbot-INL) [reply tests/framework/Optimizers/GradientDescent/gold/Basic/opt_export.xml:5]: this file tests the output of the data from an optimization, and is already tested as part of the regression system (see `tests`.

File: framework/Optimizers/RavenSampled.py (hunk 2 above)
Comment (PaulTalbot-INL) [reply framework/Optimizers/RavenSampled.py:610]: sure, moving.

File: tests/framework/AnalyticModels/optimizing/beale.py (hunk 3 above)
Comment (wangcj05) [reply tests/framework/AnalyticModels/optimizing/beale.py:39]: This is my suggestion, we better follow the standards when we want to push any codes into our repository.

File: tests/framework/Optimizers/GradientDescent/gold/Basic/opt_export.xml (hunk 5 above)
Comment (wangcj05) [reply tests/framework/Optimizers/GradientDescent/gold/Basic/opt_export.xml:5]: These variables are also included in the output csv file which are tested. It seems to me this file is not needed to be tested.

**Review by wangcj05 — APPROVED** [review]

**Pull request merged**